Topics/Software_Engineering/Retries.md

```diff
@@ -0,0 +1,86 @@
+# Retries in web services
+
+Especially relevant for webserver applications, but useful for others, retries are really tricky to get right. 
+Retries and throttling are both terms used to talk about the _flow_ of traffic into a service. Often the 
+operators/developers of that service want to make guarantees about the rate of that flow, or otherwise direct 
+traffic.
+
+## Why retry at all?
+
+Intermittent failure can happen at any level. This can be within a single host if its on-host disk, memory, or 
+CPU fails, or often in the communication between two hosts over a network. Networks, especially over the public 
+internet using TCP/IP, are known to have periodic failures due to high load/congestion or network infrastructure 
+hardware failure.
+
+Retries are a really simple, easy answer to these intermittent failures. If the error only happens rarely, then 
+trying a task again is a really effective way to ensure that the message goes through. This often manifests 
+itself as retrying API calls.
+
+Some services have built up language around these retries to control them. For example, calling AWS APIs returns
+metadata about the request itself. For example, there is a `$retryable` field in most of the AWS SDKv2's APIs, 
+the most common client used to make AWS API calls. If this field is set `true`, the server is hinting that the 
+failure was intermittent and that the client should retry. If the field is set to `false`, the server is hinting 
+to the client that the failure is likely going to happen again.
+
+## What are the problems with retries?
+
+Since retries are so simple to implement and elegant, they are usually the first tool that developers reach for 
+when a dependency of theirs has intermittent failures, but how can this go wrong?
+
+Consider a case where 4 distinct software teams each build products that depend on one another, in a chain like:
+```
+A -> B -> C -> D
```

Comment:
I like how you gave an example that's easy and clear to follow!I like how you gave an example that's easy and clear to follow!

Reply:
I agree, the example added a lot of clarity to how retries can work in practice and how using it in these kinds of situations can lead to problems.